### PR TITLE
[ALL CHARTS] Publish new charts to OCI registry

### PR DESCRIPTION
**Pull request by scottrigby**
Signed-off-by: Scott Rigby <scott@r6by.com>

#### What this PR does / why we need it

GH Action to provide Helm OCI support for existing and new chart versions.

#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- partially addresses #2454 

#### Special notes for your reviewer

- You can see a working example here: https://github.com/scottrigby/ghcr-helm-oci
- [ ] Note pushing previous chart versions to OCI will be a separate step.
- [ ] Also update readme telling people how to verify artifacts with cosign

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- ~~[ ] Chart Version bumped~~
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -37,3 +39,32 @@ jobs:
...
+          for pkg in .cr-release-packages/*; do
+            if [ -z "${pkg:-}" ]; then
+              break
+            fi
+            helm push "${pkg}" oci://ghcr.io/${GITHUB_REPOSITORY_OWNER}/charts
+            file=${pkg##*/}
+            name=${file%-*}
+            version=${file%.*}
+            version=${version#*-}
+            cosign sign ghcr.io/${GITHUB_REPOSITORY_OWNER}/charts/"${name}":"${version}"
```
Comment (torstenwalter) [inline .github/workflows/release.yaml:69]: Should we quote everything?
```suggestion
            cosign sign "ghcr.io/${GITHUB_REPOSITORY_OWNER}/charts/${name}:${version}"
```

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -37,3 +39,32 @@ jobs:
...
+        # when filling gaps with previously released charts, cr would create
+        # nothing in .cr-release-packages/, and the original globbing character
+        # would be preserved, causing a non-zero exit. Set nullglob to fix this
+        run: |
+          shopt -s nullglob
+          for pkg in .cr-release-packages/*; do
+            if [ -z "${pkg:-}" ]; then
+              break
+            fi
+            helm push "${pkg}" oci://ghcr.io/${GITHUB_REPOSITORY_OWNER}/charts
```
Comment (torstenwalter) [inline .github/workflows/release.yaml:64]:
```suggestion
            helm push "${pkg}" "oci://ghcr.io/${GITHUB_REPOSITORY_OWNER}/charts"
```

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -37,3 +39,32 @@ jobs:
...
+            file=${pkg##*/}
+            name=${file%-*}
```
Comment (torstenwalter) [inline .github/workflows/release.yaml:66]: Do we need quotes here as well? Could $file contain spaces?

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -37,3 +39,32 @@ jobs:
...
+
+      # see https://github.com/helm/chart-releaser/issues/183
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - uses: sigstore/cosign-installer@main
```
Comment (monotek) [inline .github/workflows/release.yaml:51]: can we use a fixed version?
```suggestion
      - uses: sigstore/cosign-installer@v2.8.1
```

**Review by monotek — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yaml (hunk 1 above)
Comment (scottrigby) [reply .github/workflows/release.yaml]: Thanks, added a single commit with you as co-author. Didn't wrap entire line, just vars

File: .github/workflows/release.yaml (hunk 2 above)
Comment (scottrigby) [reply .github/workflows/release.yaml]: Thanks! added a single commit with you as co-author 16711033bf9a007cb75221d12fb5e1689f0685ba. Didn't wrap entire line, just vars

File: .github/workflows/release.yaml (hunk 1 above)
Comment (scottrigby) [reply .github/workflows/release.yaml]: We don't need to quote the bash var expansion syntax. I ran actionlint locally, and shellcheck (which it uses to check this part) – only those lines needed love 😄

File: .github/workflows/release.yaml (hunk 3 above)
Comment (scottrigby) [reply .github/workflows/release.yaml:66]: We don't need to quote the bash var expansion syntax. I ran actionlint locally, and shellcheck (which it uses to check this part) – only those lines needed love 😄

**Review by monotek — APPROVED** [review]

**Comment by scottrigby** [comment]
> Oh I want to do one more thing, add info to the readme on how to pull from ghcr, and how to verify keyless signature using cosign...

Let's do this as a follow-up PR, once we see everything working in ghcr.io.

PS here's what I was thinking as a template for updating each charts README's. We also prob want to add a single note in the main repo README as well.

```diff
diff --git a/charts/prometheus/README.md b/charts/prometheus/README.md
index d8a1e9ab2..055bc61be 100644
--- a/charts/prometheus/README.md
+++ b/charts/prometheus/README.md
@@ -9,24 +9,41 @@ This chart bootstraps a [Prometheus](https://prometheus.io/) deployment on a [Ku
 - Kubernetes 1.16+
 - Helm 3+
 
-## Get Repo Info
+## Install from ghcr.io (OCI)
 
 ```console
-helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
-helm repo update
+helm install [RELEASE_NAME] \
+oci://ghcr.io/prometheus-community/charts/prometheus
+```
+
+_See [Using OCI-based registries](https://helm.sh/docs/topics/registries/) for documentation._
+
+## Verify the OCI artifact
+
+Install the [cosign](https://github.com/sigstore/cosign) CLI:
+
+```sh
+brew install sigstore/tap/cosign
/ ```
 
-_See [helm repo](https://helm.sh/docs/helm/helm_repo/) for command documentation._
+Verify a prometheus release with cosign CLI:
+
+```sh
+COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/prometheus-community/charts/prometheus:[VERSION]
+```
 
-## Install Chart
+## Install from github.io (HTTP)
 
 ```console
+helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
+helm repo update
+
 helm install [RELEASE_NAME] prometheus-community/prometheus
/ ```
 
 _See [configuration](#configuration) below._
 
-_See [helm install](https://helm.sh/docs/helm/helm_install/) for command documentation._
+_See [helm repo](https://helm.sh/docs/helm/helm_repo/) and [helm install](https://helm.sh/docs/helm/helm_install/) for command documentation._
 
 ## Dependencies
 

```

**Review by torstenwalter — APPROVED** [review]

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -37,3 +39,32 @@ jobs:
...
+          for pkg in .cr-release-packages/*; do
+            if [ -z "${pkg:-}" ]; then
+              break
+            fi
+            helm push "${pkg}" oci://ghcr.io/"${GITHUB_REPOSITORY_OWNER}"/charts
+            file=${pkg##*/}
+            name=${file%-*}
+            version=${file%.*}
+            version=${version#*-}
+            cosign sign ghcr.io/"${GITHUB_REPOSITORY_OWNER}"/charts/"${name}":"${version}"
```
Comment (developer-guy) [inline .github/workflows/release.yaml:69]: Hi @scottrigby, instead of signing the tags, it'd be better to use digests to provide a better assurance about the content we are signing. I did a similar task in the Kyverno project, as you can see from [here](https://github.com/kyverno/kyverno/blob/5557d6c85d7e1866ebd3bed5c1c3a4c9d55a3881/.github/workflows/helm-release.yaml#L93). 

**TLDR**;

_AFAIK, the `helm push` command outputs the digest of the image being pushed, so we can save that digest in a file and retrieve it from there later._

WDYT?

Comment (developer-guy) [reply .github/workflows/release.yaml:69]: Kindly ping @monotek @torstenwalter, do you have any comments about it?

Comment (developer-guy) [reply .github/workflows/release.yaml:69]: _https://github.com/prometheus-community/helm-charts/pull/2664_
